palinternal.h breaks "errno": src/pal/src/include/pal/thread.hpp:498:13: error: use of undeclared identifier 'errno'

I was [unable to compile CoreCLR on FreeBSD](https://github.com/dotnet/coreclr/issues/1044#issuecomment-107916239), due to `errno` being apparently undefined at the point where it's needed in `src/pal/src/include/thread.hpp`. ([full log](https://gist.github.com/saper/eafc21d80b753511270d)).

The problem might have been uncovered (but not caused) by f1b2b7d934339b0def5a96f111ea954974156ff8, which (re)introduced the use of system `errno` instead of using some own implementation.

The problem is that `palinternal.h` is a clever trick that does not handle some system headers fully.
It is worth to read https://github.com/dotnet/coreclr/blob/master/src/pal/src/include/pal/palinternal.h#L14-L136 before proceeding :)

What happens is this:

- `src/exception/seh.cpp` includes `<typeinfo>` and then `"pal/thread.hpp"`
- on my system `<typeinfo>` includes `<errno.h>`
- `<palinternal.h>` does its funny tricks and together with `<pal.h>` try to redefine `errno`. 
- What happens in the end is that `errno` stays undefined once the `corunix.hpp` is processed.
- Subsequent `#include <errno.h>` in the `thread.hpp` DOES NOT help, because `<errno.h>` was already included and the header guard (in my case `_SYS_ERRNO_H_`) is already defined
- This way we never get original system macro definition of `errno` back.
- This problem does not occur in the other files because at the point `palinternal.h` gets used `<errno.h>` has not yet been included by any other file.

So there needs to be a warning in the 'palinternal.h' file that explains the problem - 
using header files of the undef'd identifiers is probably a no-no.

When moving some function from category `A.2` (as described in `palinternal.h`) to category `A.1` 
it needs to be carefully checked if everything gets defined again.

Not sure where the bug is, maybe `#undef errno` and friends should go away, maybe `errno` should not be redefined in `pal.h`... etc. etc.

Anyway, my workaround was to include `<typeinfo>` *after* `"pal/thread.hpp"` in the `seh.cpp`
and the CoreCLR got compiled; but that is certainly not a solution.

This problem is not very FreeBSD specific, any use of some system header before `palinternal.h` gets included may create such problems, or worse, hidden bugs.

(mentioning https://github.com/dotnet/coreclr/issues/1010)

Thanks @josteink @jasonwilliams200OK for assisting me to get this sorted out.

We build with skipmscorlib, so mono would not be involved.

Sent from my Windows Phone
________________________________
From: Peter Jas<mailto:notifications@github.com>
Sent: ‎10/‎7/‎2015 10:30 PM
To: dotnet/coreclr<mailto:coreclr@noreply.github.com>
Cc: Matt Mitchell<mailto:mmitche@microsoft.com>
Subject: Re: [coreclr] palinternal.h breaks "errno": src/pal/src/include/pal/thread.hpp:498:13: error: use of undeclared identifier 'errno' (#1091)


@mmitche<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fmmitche&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=qiIR05TPEFn8hgsYi9xa0WAcPOH4P3gtS%2bFKPd4b%2bTs%3d>, with the VM recipe for CoreCLR (https://gist.github.com/jasonwilliams200OK/06db121ec534821fe541<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgist.github.com%2fjasonwilliams200OK%2f06db121ec534821fe541&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=8k7BaOJ8B2C7H6SMit1XOC8oEOwCW1ovq2x5WpR3pDc%3d>), I am able to successfully build locally on FreeBSD 10.2 by applying @saper<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fsaper&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=Ouzo5WHATbHRUV19ysxIcDlmqLPSJE%2bYrOz6atjFylY%3d>'s patch<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fsaper%2fcoreclr%2fcommit%2f9deed7c6b8b85de418362c258b59d5d6fa56046f&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=9lop9UF3JUikaJ8sLPG27Al5wZjVzvPxCkFM%2f5VugLA%3d>:

diff --git a/src/pal/src/exception/seh.cpp b/src/pal/src/exception/seh.cpp
index 3080814..812d909 100644
--- a/src/pal/src/exception/seh.cpp
+++ b/src/pal/src/exception/seh.cpp
@@ -19,8 +19,8 @@ Abstract:

 --*/

-#include <typeinfo>
 #include "pal/thread.hpp"
+#include <typeinfo>
 #include "signal.hpp"
 #include "pal/handleapi.hpp"
 #include "pal/seh.hpp"

I wonder how it is building on CI, if CI is running FreeBSD 10.[0-2] and mono 4.0.3? Because locally without applying this patch and setenv TZ UTC ; setenv MONO_THREADS_PER_CPU 50 before ./build.sh it does not succeed with current CoreCLR master.

Should we fix this issue with the said patch?

/cc: @jkotas<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fjkotas&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=lZCGIqBIh2Z7A57Y2LDuggGUH%2byjc415%2fB234HWujFI%3d>

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcoreclr%2fissues%2f1091%23issuecomment-146424492&data=01%7c01%7cmmitche%40microsoft.com%7cb4515c3ecddc41600bf208d2cfa188ee%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=202pMQ5q9EDTFZyM2fcoCEk6wHpxGAd1wbZl1rSubfg%3d>.

I'll try and find out why its working in CI tomorrow.

Sent from my Windows Phone
________________________________
From: Peter Jas<mailto:notifications@github.com>
Sent: ‎10/‎7/‎2015 10:51 PM
To: dotnet/coreclr<mailto:coreclr@noreply.github.com>
Cc: Matt Mitchell<mailto:mmitche@microsoft.com>
Subject: Re: [coreclr] palinternal.h breaks "errno": src/pal/src/include/pal/thread.hpp:498:13: error: use of undeclared identifier 'errno' (#1091)


You are right! I missed that one and will try with skipmscorlib.
Nonetheless the code patch is still required to compile the code. :)

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcoreclr%2fissues%2f1091%23issuecomment-146427469&data=01%7c01%7cmmitche%40microsoft.com%7c36c32193ee6c45a61ce808d2cfa47546%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=kx8aARSpfzSaMuex70gdljSx1H2q3Wou8X0SGh9JFs0%3d>.

--- a/src/pal/src/exception/seh.cpp
+++ b/src/pal/src/exception/seh.cpp
@@ -19,9 +19,7 @@ Abstract:
 
 --*/
 
-#include <typeinfo>
 #include "pal/thread.hpp"
-#include "signal.hpp"
 #include "pal/handleapi.hpp"
 #include "pal/seh.hpp"
 #include "pal/dbgmsg.h"
@@ -30,18 +28,19 @@ Abstract:
 #include "pal/init.h"
 #include "pal/process.h"
 #include "pal/malloc.hpp"
+#include "signal.hpp"
 
 #if HAVE_ALLOCA_H
 #include "alloca.h"
 #endif
 
-#include <errno.h>
-#include <string.h>
 #if HAVE_MACH_EXCEPTIONS
 #include "machexception.h"
 #else
 #include <signal.h>
 #endif
+
+#include <string.h>
 #include <unistd.h>
 #include <pthread.h>
 #include <stdlib.h>

